src/operator/tensor/elemwise_unary_op.h

```diff
@@ -31,173 +31,337 @@
-  using namespace mxnet_op;
-  Stream<xpu> *s = ctx.get_stream<xpu>();
 
-  CHECK_EQ(inputs.size(), 1U);
-  CHECK_EQ(outputs.size(), 1U);
-  MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
-    Kernel<op, xpu>::Launch(s, outputs[0].Size(),
-      outputs[0].dptr<DType>(), inputs[0].dptr<DType>());
-  });
-}
+class OpBase {
+ public:
+  template<int Req>
+  struct MapSetToScalar {
+    template<typename DType>
+    MSHADOW_XINLINE static void Map(int i, DType *out, const DType value) {
+      KERNEL_ASSIGN(out[i], Req, value);
+    }
+  };
 
-template<typename GRAD_OP>
-struct unary_bwd {
-  template<typename DType>
-  MSHADOW_XINLINE static DType Map(DType a, DType b) {
-    return DType(a*GRAD_OP::Map(b));
+ protected:
+  /*! \brief Copy blob data */
+  template<typename xpu>
+  static void inline CopyBlob(mshadow::Stream<xpu> *s,
+                              const TBlob *dest_blob,
+                              const OpReqType reqi,
+                              const TBlob& src_blob) {
+    CHECK_EQ(src_blob.type_flag_, dest_blob->type_flag_);
+    CHECK_EQ(src_blob.shape_, dest_blob->shape_);
+    MSHADOW_TYPE_SWITCH(src_blob.type_flag_, DType, {
+      // Check if the pointers are the same (in-place operation needs no copy)
+      if (reqi != kNullOp && src_blob.dptr<DType>() != dest_blob->dptr<DType>()) {
+        mshadow::Copy(dest_blob->FlatTo1D<xpu, DType>(s), src_blob.FlatTo1D<xpu, DType>(s), s);
+      }
+    });
   }
-};
 
-template<typename xpu, typename OP>
-void UnaryCompute(const nnvm::NodeAttrs& attrs,
-                  const OpContext& ctx,
-                  const std::vector<TBlob>& inputs,
-                  const std::vector<OpReqType>& req,
-                  const std::vector<TBlob>& outputs) {
-  using namespace mshadow;
-  using namespace mshadow::expr;
-  Stream<xpu> *s = ctx.get_stream<xpu>();
-  MSHADOW_TYPE_SWITCH(outputs[0].type_flag_, DType, {
-    Tensor<xpu, 1, DType> out = outputs[0].FlatTo1D<xpu, DType>(s);
-    ASSIGN_DISPATCH(out, req[0], F<OP>(inputs[0].FlatTo1D<xpu, DType>(s)));
-  });
-}
+  /*! \brief Allocate geometry-related blob data for sparse tensors
+   * \param dest Destination sparse NDArray
+   * \param clone_from sparse NDArray from which to clone storage attributes
+   */
+  static void AllocateGeometry(const NDArray *dest, const NDArray* clone_from = nullptr) {
+    if (clone_from) {
+      const TShape ishape = clone_from->storage_shape();
+      TShape sshape = dest->storage_shape();
+      CHECK(shape_assign(&sshape, ishape));
+      dest->CheckAndAllocData(sshape);
+      CHECK_EQ(dest->storage_type(), clone_from->storage_type());
+      for (size_t i = 0, n = clone_from->aux_shapes().size(); i < n; ++i) {
+        TShape ashape = dest->aux_shape(i);
+        CHECK(shape_assign(&ashape, clone_from->aux_shape(i)));
+        dest->CheckAndAllocAuxData(i, ashape);
+      }
+      DCHECK_EQ(dest->aux_shapes().size(), clone_from->aux_shapes().size());
+    } else {
```

Comment:
I dont think infer shape will set the shape for data and aux data. When you `bind` the graph, infer shape is only called once, but you can call `forward` on the graph with different input multiple times. The actual aux shape and storage shape of the output depends on the input at runtime. That means you need to defer setting the shape for aux data during infershape, and instead doing that inside the operator at run time. So I don't think `InferShape` is supposed to affect aux_shape and storage_shape at all.